### ISSUE TITLE
Interpolated IK functionality

### ISSUE THREAD

**Issue by mateiciocarlie** [moveit/moveit_ros#127]
Add InterpolatedIK functionality. Note that this is needed for both pick&place (the grasping pipeline) and for benchmarking; it would make sense for the same functionality to be used in both.

There has already been a lot of work on this and the component is already done, just adding the ticket so we can discuss any remaining issues, keep track of it, and depend on it in other tickets.

@isucan : anything in particular that is still an interesting discussion topic here? Or can this be closed directly?

**Comment by isucan** [comment]
This exists now, but I am not entirely convinced it will work for all our use cases. I think that will generate different tickets though, so closing for now.

**Comment by mateiciocarlie** [comment]
Could you please point me to the API for it? Presumably it's a C++ API? Thanks!

**Comment by isucan** [comment]
The input is actually specified as a planning request:
https://github.com/ros-planning/moveit_msgs/blob/groovy-devel/msg/MotionPlanRequest.msg
The part that corresponds to interpolated IK is specified as a sequence of constraints to be achieved in order, via straight lines, using this message:
https://github.com/ros-planning/moveit_msgs/blob/groovy-devel/msg/TrajectoryConstraints.msg

This functionality can be used via the normal planning interface (move_group) at the ROS level. At C++ level, there are multiple points of entry:
- direct call to the planning interface plugin (the service request contains an instance of the MotionPlanRequest message)
  https://github.com/ros-planning/moveit_core/blob/groovy-devel/planning_interface/include/moveit/planning_interface/planning_interface.h
- call to a planning_pipeline (which also calls planning request adapters as needed)
  https://github.com/ros-planning/moveit_ros/blob/groovy-devel/planning/planning_pipeline/include/moveit/planning_pipeline/planning_pipeline.h
- call to an execution pipeline (which also executes and monitors the solution path)
  https://github.com/ros-planning/moveit_ros/blob/groovy-devel/planning/plan_execution/include/moveit/plan_execution/plan_execution.h
- call to the move_group_interface (simplest interface)
  https://github.com/ros-planning/moveit_ros/blob/groovy-devel/planning_interface/move_group_interface/include/moveit/move_group_interface/move_group.h
  (one would call followConstraints() here and then either plan() or move(); this functionality builds on top of the previous two libraries)

**Comment by mateiciocarlie** [comment]
After looking through these API's, I was thinking how some of the needs of the grasping pipeline for InterpolatedIK could be implemented. In particular:

More important:
- when planning from pre-grasp to grasp, neither is specified as a joint state. The pre-grasp does not pre-suppose a particular joint config, rather, it is a Cartesian end-effector pose just like the grasp. In this sense, there is not a StartState as the MotionPlanningRequest expects it. 

Maybe we can do without:
- the length of the InterpolatedIK trajectory from pre-grasps to grasps is not pre-set, but rather determined at run-time. Essentially, InterpolatedIK is asked to find a trajectory as long as possible up to X (usually =10) cm. If the found trajectory is at least Y (=5) cm long, then the grasp is executed. This basically says "ideally I would like the path from pre-grasp to grasp to be 10cm but am willing to execute the grasp if it's at least 5". I am not sure how this can be implemented with the available API. This feature was very important with our old implementation, given that InterpolatedIK was very brittle. If this new version is more robust, maybe we can get away with a hard setting on how long the path should be.

Note that everything mentioned there about pre-grasp -> grasp also applies to the other InterpolatedIK trajectories computed during pick&place:
- grasps -> lift
- pre-place -> place
- place -> retreat

Should these go in separate tickets, or maybe re-open this one?

**Comment by mateiciocarlie** [comment]
Oh, sorry for the unclear parts. Let me try to expand:

Basically, yes - InterpolatedIK does not have a "start state" with joint angles - hence the "IK" part. We just have a gripper path along a straight line in Cartesian space, and are trying to find joint angles for all of it.

Only after the traditional InterpolatedIK was complete, we would run a motion plan request from the robot's start state (its current state), and the goal would be whatever joint angles InterpolatedIK returned for the start of the interpolated trajectory.

If we keep the motion plan to the pre-grasp and the InterpolatedIK from pre-grasp to grasp separate, I do not think there is a way around it - essentially, the start state for InterpolatedIK is just a gripper pose, or more precisely a gripper pose constraint.

**Comment by isucan** [comment]
Ah, ok. I understand it now. This is not exactly how things are running now for the pick & place. I'll think about this some more.

**Comment by isucan** [comment]
I believe this is now done in https://github.com/ros-planning/moveit_core/blob/groovy-devel/kinematic_state/include/moveit/kinematic_state/joint_state_group.h
